### PR TITLE
feat(code): add options "create-bytes" and "update-bytes"

### PR DESCRIPTION
**Pull request by thwyr**
Add feature flags for reference- and cow- types for byte blobs in update/create akin to the feature flags for string types.

This allows one to avoid additional allocations or moves in Create and Update structs when dealing with byte blobs similar to allocations avoided by using &str and Cow<str> with strings.

### CODE REVIEW

File: src/bin/main.rs [hunk 1]
```diff
@@ -140,6 +148,27 @@ impl From<StringTypeCli> for StringType {
...
+#[derive(Debug, ValueEnum, Clone, PartialEq, Default)]
+pub enum BytesTypeCli {
+    /// Use "String"
+    #[default]
+    Vec,
+    /// Use "&str"
+    Slice,
+    /// Use "Cow<str>"
+    Cow,
+}
```
Comment (hasezoey) [inline src/bin/main.rs:160]: the comments should likely be updated

File: src/code.rs [hunk 2]
```diff
@@ -314,6 +314,12 @@ impl<'a> Struct<'a> {
                     StructType::Update => self.opts.get_update_str_type().as_str().to_string(),
                     StructType::Create => self.opts.get_create_str_type().as_str().to_string(),
                 }
+            } else if f.base_type == "Vec<u8>" {
+                f.base_type = match self.ty {
+                    StructType::Read => f.base_type,
+                    StructType::Update => self.opts.get_update_bytes_type().as_str().to_string(),
+                    StructType::Create => self.opts.get_create_bytes_type().as_str().to_string(),
+                }
```
Comment (hasezoey) [inline src/code.rs:322]: in this file the `lifetimes` also will need to be updated, at https://github.com/Wulf/dsync/blob/3f3d57c36647451403c907c6078764fa3894463d/src/code.rs#L293-L298

File: test/create_update_bytes_cow/models/todos/generated.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    pub data_nullable: Option<Cow<'a, [u8]>>,
+}
+
+/// Update Struct for a row in table `todos` for [`Todos`]
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, AsChangeset, PartialEq, Default)]
+#[diesel(table_name=todos)]
+pub struct UpdateTodos {
+    /// Field representing column `data_nullable`
+    pub data_nullable: Option<Option<Cow<'a, [u8]>>>,
+}
```
Comment (hasezoey) [inline test/create_update_bytes_cow/models/todos/generated.rs:35]: lifetimes are currently not accounted for

**Review by hasezoey — CHANGES_REQUESTED** [review]
mostly looks good, though there are some points:
- the `main.rs` struct comments will need to be updated (are still copy-paste from the string version)
- the lifetimes are not accounted for yet (only string lifetimes, see comment on `code.rs`)
- also consider adding a test for a lifetime string & lifetime bytes test combined (where both are set to generate lifetimes like both being `cow` or `slice`)

also dont just post a PR without a description, provide a simple explanation of what this PR does and why it may be necessary (instead of just using `Vec<[u8]>` everywhere)

**Comment by thwyr** [comment]
I should have fixed up everything at this point; I apologize for the half-assed nature of the initial PR.

Looking at the code now, I'm convinced that there is a compelling argument to merge the two into a single flag; I struggle to think of a situation where the user wants to avoid moves/allocations for Strings but not for Vec<u8>.

**Review by hasezoey — APPROVED** [review]
This PR looks good now, thanks for the work (i at least dont see any issues, but the test output is currently not compile-tested, see #114)

> I'm convinced that there is a compelling argument to merge the two into a single flag; I struggle to think of a situation where the user wants to avoid moves/allocations for Strings but not for Vec.

i guess you could say that, but this is for the future and if changed, then likely will only be changed for the `main.rs` configuration (because why change the library one, if it already exists and is not much burden and so has more configurability?)

**Comment by thwyr** [comment]
> This PR looks good now, thanks for the work (i at least dont see any issues, but the test output is currently not compile-tested, see #114)

Thanks. It looks like the compile-testing is a WIP. I can revisit adding that onto this once I have a merged example to work with.


> i guess you could say that, but this is for the future and if changed, then likely will only be changed for the `main.rs` configuration (because why change the library one, if it already exists and is not much burden and so has more configurability?)

My main motivation would be to simplify the library code; there is something to be said for configurability being meaningless if it's not actually put to use.

Having said that, I've already scratched my itch. I also have a follow-up commit documenting the flags.

**Pull request merged**